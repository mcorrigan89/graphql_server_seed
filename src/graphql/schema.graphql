enum CacheControlScope {
  PUBLIC
  PRIVATE
}

directive @cacheControl (
  maxAge: Int
  scope: CacheControlScope
) on FIELD_DEFINITION | OBJECT | INTERFACE


type Query {
  _: String
}

type Mutation {
  _: String
}

type Subscription {
  _: String
}