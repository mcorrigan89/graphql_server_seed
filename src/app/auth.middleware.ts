import { Request, RequestHandler } from 'express';
import { verifyToken } from './token';

export const getToken = (req: Request) => req.headers.authorization;

export interface DecodedAuthToken {
  id: string;
}
export interface AuthenticatedRequest extends Request {
  decodedToken?: DecodedAuthToken;
}

export const tokenCheck: RequestHandler = async (req: AuthenticatedRequest, _, next) => {
  const token = getToken(req);
  if (token) {
    try {
      const decodedToken = verifyToken<DecodedAuthToken>(token);
      req.decodedToken = decodedToken;
      next();
    } catch (err) {
      console.error(err);
      next();
    }
  } else {
    next();
  }
};
