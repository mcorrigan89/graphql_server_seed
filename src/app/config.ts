/* tslint:disable */
/*
  "In particular, if there is a variable in your .env file
  which collides with one that already exists in your environment,
  then that variable will be skipped."
  - https://www.npmjs.com/package/dotenv
*/

import { ConnectionOptions } from 'typeorm';
import * as dotenv from 'dotenv';
dotenv.config();
import { MODELS } from '@data/index';

const env = process.env;
export const isProduction: boolean = env.NODE_ENV === 'production';
export const isTest: boolean = env.NODE_ENV === 'test' || env.NODE_ENV === 'ci';

export interface DatabaseConfig {
  host: string;
  database: string;
  userName?: string;
  password?: string;
}

export interface RedisConfig {
  host: string;
}
export interface ServerConfig {
  secret: string;
  port: string;
  database: DatabaseConfig;
}

export const serverConfig: ServerConfig = {
  secret: env.SECRET as string,
  port: env.PORT as string,
  database: {
    host: env.POSTGRES_HOST as string,
    database: env.POSTGRES_DATABASE as string,
    userName: env.POSTGRES_USER as string,
    password: env.POSTGRES_PASSWORD as string
  }
};

export const connectionOptions: ConnectionOptions[] = [
  {
    name: 'default',
    type: 'postgres',
    host: env.POSTGRES_HOST || '',
    port: Number(env.POSTGRES_PORT) || 5432,
    username: env.POSTGRES_USER || '',
    password: env.POSTGRES_PASSWORD || '',
    database: env.POSTGRES_DATABASE || '',
    synchronize: true,
    dropSchema: false,
    logging: false,
    cache: true,
    entities: MODELS,
    migrations: env.NODE_ENV === 'production' ? [] : ['migration/*.ts'],
    subscribers: env.NODE_ENV === 'production' ? ['dist/subscriber/**/*.js'] : ['src/subscriber/**/*.ts'],
    cli: {
      migrationsDir: 'migration'
    }
  },
  {
    name: 'test',
    type: 'postgres',
    host: env.POSTGRES_HOST || '',
    port: Number(env.POSTGRES_PORT) || 5432,
    username: env.POSTGRES_USER || '',
    password: env.POSTGRES_PASSWORD || '',
    database: env.POSTGRES_DATABASE || '',
    synchronize: true,
    dropSchema: true,
    logging: false,
    cache: true,
    entities: MODELS
  }
];
