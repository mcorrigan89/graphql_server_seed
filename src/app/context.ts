import { UserController } from '@data/user/controller';
import { UserView } from '@data/user/view';

export class Context {
  public userController: UserController;
  private currentUser: UserView | null = null;

  constructor() {
    this.userController = new UserController(this);
  }

  public getCurrentUser = () => {
    return this.currentUser;
  };

  public setCurrentUser = async (id: string) => {
    this.currentUser = await this.userController.getUserById(id);
  };
}
