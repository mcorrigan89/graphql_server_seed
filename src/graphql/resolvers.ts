import { Resolvers } from '@graphql/resolver.types';

export const resolvers = (): Resolvers => ({
  Query: {
    me: (_, __, context) => context.getCurrentUser(),
    user: (_, args, context) => context.userController.getUserById(args.id),
    users: (_, __, context, info) => context.userController.getUsers()
  },
  Mutation: {
    createUser: (_, args, context) => context.userController.createUser(args.payload),
    login: (_, args, context) => context.userController.login(args.payload)
  }
});
