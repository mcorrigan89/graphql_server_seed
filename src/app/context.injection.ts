import { UserController } from '@data/user/controller';
import { Context } from './context';
import { container, TYPES } from './context.setup';

container.bind<Context>(TYPES.CONTEXT).to(Context);
container.bind<UserController>(TYPES.USER_CONTROLLER).to(UserController);

const injectedContainer = container;

export { injectedContainer };
