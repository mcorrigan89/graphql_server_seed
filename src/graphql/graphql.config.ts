import { ApolloServer } from 'apollo-server-express';
import { Context } from '@app/context';
import { AuthenticatedRequest } from '@app/auth.middleware';
import { merge } from 'lodash';
import { addResolveFunctionsToSchema } from 'apollo-server';
import { KeyValueCache } from 'apollo-server-caching';
import { resolvers } from '@graphql/resolvers';
import { Server } from 'http';
import { loadSchema } from '@graphql-tools/load';
import { GraphQLFileLoader } from '@graphql-tools/graphql-file-loader';
import responseCachePlugin from 'apollo-server-plugin-response-cache';

export const createSchema = async () => {
  const schemaTypeDefs = await loadSchema('./src/**/*.graphql', { loaders: [new GraphQLFileLoader()] });
  const schema = addResolveFunctionsToSchema({
    schema: schemaTypeDefs,
    resolvers: merge(resolvers())
  });
  return schema;
};

export const createApolloServer = async (server: Server) => {
  const schema = await createSchema();
  const apolloServer = new ApolloServer({
    schema,
    introspection: true,
    plugins:[responseCachePlugin({
      sessionId: (requestContext) => { 
        const context = requestContext.context as Context;
        const user = context.getCurrentUser();
        const userId = user?.id;
        return userId ?? null;
      },
    })],
    context: async ctx => {
      try {
        const context = new Context();
        const token = (ctx.req as AuthenticatedRequest).decodedToken;
        if (token) {
          await context.setCurrentUser(token.id);
        }
        return context;
      } catch (err) {
        console.log(err);
      }
    }
  });
  return { apolloServer, schema };
};
