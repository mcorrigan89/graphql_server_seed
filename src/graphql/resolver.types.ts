import { GraphQLResolveInfo } from 'graphql';
import { Context } from '@app/context';
export type Maybe<T> = T extends PromiseLike<infer U> ? Promise<U | null> : T | null;
export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]: Maybe<T[SubKey]> };
export type FieldWrapper<T> = T | Promise<T> | (() => T) | (() => Promise<T>);
export type EntireFieldWrapper<T> = T | Promise<T> | (() => T) | (() => Promise<T>);
export type RequireFields<T, K extends keyof T> = { [X in Exclude<keyof T, K>]?: T[X] } & { [P in K]-?: NonNullable<T[P]> };
/** All built-in and custom scalars, mapped to their actual values */
export type Scalars = {
  ID: string;
  String: string;
  Boolean: boolean;
  Int: number;
  Float: number;
  _FieldSet: any;
};






export enum CacheControlScope {
  Public = 'PUBLIC',
  Private = 'PRIVATE'
}

export type CreateUserPayload = {
  username: Scalars['String'];
  password: Scalars['String'];
  firstName?: Maybe<Scalars['String']>;
  lastName?: Maybe<Scalars['String']>;
};

export type LoginPayload = {
  username: Scalars['String'];
  password: Scalars['String'];
};

export type Mutation = {
  __typename?: 'Mutation';
  _?: EntireFieldWrapper<Maybe<FieldWrapper<Scalars['String']>>>;
  createUser: EntireFieldWrapper<FieldWrapper<User>>;
  login: EntireFieldWrapper<FieldWrapper<Scalars['String']>>;
};


export type MutationCreateUserArgs = {
  payload: CreateUserPayload;
};


export type MutationLoginArgs = {
  payload: LoginPayload;
};

export type Query = {
  __typename?: 'Query';
  _?: EntireFieldWrapper<Maybe<FieldWrapper<Scalars['String']>>>;
  me?: EntireFieldWrapper<Maybe<FieldWrapper<User>>>;
  user: EntireFieldWrapper<FieldWrapper<User>>;
  users: EntireFieldWrapper<Array<FieldWrapper<User>>>;
};


export type QueryUserArgs = {
  id: Scalars['ID'];
};

export type Subscription = {
  __typename?: 'Subscription';
  _?: EntireFieldWrapper<Maybe<FieldWrapper<Scalars['String']>>>;
};

export type User = {
  __typename?: 'User';
  id: EntireFieldWrapper<FieldWrapper<Scalars['ID']>>;
  username: EntireFieldWrapper<FieldWrapper<Scalars['String']>>;
  firstName?: EntireFieldWrapper<Maybe<FieldWrapper<Scalars['String']>>>;
  lastName?: EntireFieldWrapper<Maybe<FieldWrapper<Scalars['String']>>>;
};


export type WithIndex<TObject> = TObject & Record<string, any>;
export type ResolversObject<TObject> = WithIndex<TObject>;

export type ResolverTypeWrapper<T> = Promise<T> | T;


export type LegacyStitchingResolver<TResult, TParent, TContext, TArgs> = {
  fragment: string;
  resolve: ResolverFn<TResult, TParent, TContext, TArgs>;
};

export type NewStitchingResolver<TResult, TParent, TContext, TArgs> = {
  selectionSet: string;
  resolve: ResolverFn<TResult, TParent, TContext, TArgs>;
};
export type StitchingResolver<TResult, TParent, TContext, TArgs> = LegacyStitchingResolver<TResult, TParent, TContext, TArgs> | NewStitchingResolver<TResult, TParent, TContext, TArgs>;
export type Resolver<TResult, TParent = {}, TContext = {}, TArgs = {}> =
  | ResolverFn<TResult, TParent, TContext, TArgs>
  | StitchingResolver<TResult, TParent, TContext, TArgs>;

export type ResolverFn<TResult, TParent, TContext, TArgs> = (
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => Promise<TResult> | TResult;

export type SubscriptionSubscribeFn<TResult, TParent, TContext, TArgs> = (
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => AsyncIterator<TResult> | Promise<AsyncIterator<TResult>>;

export type SubscriptionResolveFn<TResult, TParent, TContext, TArgs> = (
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => TResult | Promise<TResult>;

export interface SubscriptionSubscriberObject<TResult, TKey extends string, TParent, TContext, TArgs> {
  subscribe: SubscriptionSubscribeFn<{ [key in TKey]: TResult }, TParent, TContext, TArgs>;
  resolve?: SubscriptionResolveFn<TResult, { [key in TKey]: TResult }, TContext, TArgs>;
}

export interface SubscriptionResolverObject<TResult, TParent, TContext, TArgs> {
  subscribe: SubscriptionSubscribeFn<any, TParent, TContext, TArgs>;
  resolve: SubscriptionResolveFn<TResult, any, TContext, TArgs>;
}

export type SubscriptionObject<TResult, TKey extends string, TParent, TContext, TArgs> =
  | SubscriptionSubscriberObject<TResult, TKey, TParent, TContext, TArgs>
  | SubscriptionResolverObject<TResult, TParent, TContext, TArgs>;

export type SubscriptionResolver<TResult, TKey extends string, TParent = {}, TContext = {}, TArgs = {}> =
  | ((...args: any[]) => SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>)
  | SubscriptionObject<TResult, TKey, TParent, TContext, TArgs>;

export type TypeResolveFn<TTypes, TParent = {}, TContext = {}> = (
  parent: TParent,
  context: TContext,
  info: GraphQLResolveInfo
) => Maybe<TTypes> | Promise<Maybe<TTypes>>;

export type IsTypeOfResolverFn<T = {}, TContext = {}> = (obj: T, context: TContext, info: GraphQLResolveInfo) => boolean | Promise<boolean>;

export type NextResolverFn<T> = () => Promise<T>;

export type DirectiveResolverFn<TResult = {}, TParent = {}, TContext = {}, TArgs = {}> = (
  next: NextResolverFn<TResult>,
  parent: TParent,
  args: TArgs,
  context: TContext,
  info: GraphQLResolveInfo
) => TResult | Promise<TResult>;

/** Mapping between all available schema types and the resolvers types */
export type ResolversTypes = ResolversObject<{
  CacheControlScope: CacheControlScope;
  CreateUserPayload: CreateUserPayload;
  String: ResolverTypeWrapper<Scalars['String']>;
  LoginPayload: LoginPayload;
  Mutation: ResolverTypeWrapper<{}>;
  Query: ResolverTypeWrapper<{}>;
  ID: ResolverTypeWrapper<Scalars['ID']>;
  Subscription: ResolverTypeWrapper<{}>;
  User: ResolverTypeWrapper<User>;
  Int: ResolverTypeWrapper<Scalars['Int']>;
  Boolean: ResolverTypeWrapper<Scalars['Boolean']>;
}>;

/** Mapping between all available schema types and the resolvers parents */
export type ResolversParentTypes = ResolversObject<{
  CreateUserPayload: CreateUserPayload;
  String: Scalars['String'];
  LoginPayload: LoginPayload;
  Mutation: {};
  Query: {};
  ID: Scalars['ID'];
  Subscription: {};
  User: User;
  Int: Scalars['Int'];
  Boolean: Scalars['Boolean'];
}>;

export type CacheControlDirectiveArgs = {   maxAge?: Maybe<Scalars['Int']>;
  scope?: Maybe<CacheControlScope>; };

export type CacheControlDirectiveResolver<Result, Parent, ContextType = Context, Args = CacheControlDirectiveArgs> = DirectiveResolverFn<Result, Parent, ContextType, Args>;

export type MutationResolvers<ContextType = Context, ParentType extends ResolversParentTypes['Mutation'] = ResolversParentTypes['Mutation']> = ResolversObject<{
  _?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>;
  createUser?: Resolver<ResolversTypes['User'], ParentType, ContextType, RequireFields<MutationCreateUserArgs, 'payload'>>;
  login?: Resolver<ResolversTypes['String'], ParentType, ContextType, RequireFields<MutationLoginArgs, 'payload'>>;
}>;

export type QueryResolvers<ContextType = Context, ParentType extends ResolversParentTypes['Query'] = ResolversParentTypes['Query']> = ResolversObject<{
  _?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>;
  me?: Resolver<Maybe<ResolversTypes['User']>, ParentType, ContextType>;
  user?: Resolver<ResolversTypes['User'], ParentType, ContextType, RequireFields<QueryUserArgs, 'id'>>;
  users?: Resolver<Array<ResolversTypes['User']>, ParentType, ContextType>;
}>;

export type SubscriptionResolvers<ContextType = Context, ParentType extends ResolversParentTypes['Subscription'] = ResolversParentTypes['Subscription']> = ResolversObject<{
  _?: SubscriptionResolver<Maybe<ResolversTypes['String']>, "_", ParentType, ContextType>;
}>;

export type UserResolvers<ContextType = Context, ParentType extends ResolversParentTypes['User'] = ResolversParentTypes['User']> = ResolversObject<{
  id?: Resolver<ResolversTypes['ID'], ParentType, ContextType>;
  username?: Resolver<ResolversTypes['String'], ParentType, ContextType>;
  firstName?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>;
  lastName?: Resolver<Maybe<ResolversTypes['String']>, ParentType, ContextType>;
  __isTypeOf?: IsTypeOfResolverFn<ParentType, ContextType>;
}>;

export type Resolvers<ContextType = Context> = ResolversObject<{
  Mutation?: MutationResolvers<ContextType>;
  Query?: QueryResolvers<ContextType>;
  Subscription?: SubscriptionResolvers<ContextType>;
  User?: UserResolvers<ContextType>;
}>;


/**
 * @deprecated
 * Use "Resolvers" root object instead. If you wish to get "IResolvers", add "typesPrefix: I" to your config.
 */
export type IResolvers<ContextType = Context> = Resolvers<ContextType>;
export type DirectiveResolvers<ContextType = Context> = ResolversObject<{
  cacheControl?: CacheControlDirectiveResolver<any, any, ContextType>;
}>;


/**
 * @deprecated
 * Use "DirectiveResolvers" root object instead. If you wish to get "IDirectiveResolvers", add "typesPrefix: I" to your config.
 */
export type IDirectiveResolvers<ContextType = Context> = DirectiveResolvers<ContextType>;