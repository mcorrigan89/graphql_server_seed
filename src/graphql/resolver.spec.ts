import { createTestClient } from 'apollo-server-testing';
import { createSchema } from './graphql.config';
import { Context } from '@src/app/context';
import { ApolloServer } from 'apollo-server';

it('fetches single launch', async () => {
  //   const userAPI = new UserAPI({ store });
  //   const launchAPI = new LaunchAPI();

  // create a test server to test against, using our production typeDefs,
  // resolvers, and dataSources.
  const schema = await createSchema();
  const server = new ApolloServer({
    schema,
    context: () => new Context()
  });

  // use the test server to create a query function
  const { query } = createTestClient(server);

  // run query against the server and snapshot the output
  const res = await query({ query: '{ me { id } }', variables: { id: 1 } });
  expect(res).toMatchSnapshot();
});
