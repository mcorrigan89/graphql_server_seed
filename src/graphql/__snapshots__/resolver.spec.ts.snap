// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetches single launch 1`] = `
Object {
  "data": Object {
    "me": null,
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
